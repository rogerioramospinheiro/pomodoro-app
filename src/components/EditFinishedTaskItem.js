import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { editFinishedTask } from '../actions/tasks';

export class EditFinishedTaskItem extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            title: props.selectedTask.title,
            description: props.selectedTask.description
        };
    }
    onSubmitHandle = (e) => {
        e.preventDefault();
        const id = this.props.selectedTask.id;
        const updates = {
            title: this.state.title,
            description: this.state.description
        };
        this.props.editFinishedTask(id, updates);
        this.props.history.push('/');
    };
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState( {title} );
    };
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState( {description} );
    };
    render() {
        return (
            <div>
                <div className="timer-display">
                    <div className="timer-display__text">
                        <Link className="finished-tasks__link" to="/">Edit</Link>
                    </div>
                </div>
                <div className="task-form">
                    <form onSubmit={this.onSubmitHandle}>
                        <div className="task-form__spacebox">
                            <input type="text"
                            className="task-form__input"
                            placeholder="Title"
                            value={this.state.title}
                            onChange={this.onTitleChange}/>
                        </div>
                        <div className="task-form__spacebox">
                            <textarea
                            className="task-form__textarea"
                            placeholder="Description"
                            value={this.state.description}
                            onChange={this.onDescriptionChange}/>
                        </div>
                        <button className="task-form__save">Save</button>
                    </form>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        selectedTask: state.tasks.finished_tasks.find( (task) => task.id === props.match.params.id )
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        editFinishedTask: (id, updates) => dispatch(editFinishedTask(id, updates))
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditFinishedTaskItem);