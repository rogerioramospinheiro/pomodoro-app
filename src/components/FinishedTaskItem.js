import React from 'react';
import moment from 'moment';
import { Link } from 'react-router-dom';

const FinishedTaskItem = ( {id, title, description, start_date, end_date} ) => (
    <Link className="finished-tasks__link" to={`/edit/${id}`}>
        <div className="finished-tasks__item">
            <div>{ title }</div>
            <div>{ description }</div>
            <div>
                <span>{ moment(start_date).format('hh:mm:ss A') }</span>
                <span>-</span>
                <span>{ moment(end_date).format('hh:mm:ss A') }</span>
            </div>
        </div>
    </Link>
);

export default FinishedTaskItem;