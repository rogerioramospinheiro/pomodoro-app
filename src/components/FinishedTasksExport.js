import React from 'react';
import { CSVLink } from 'react-csv';
import moment from 'moment';

const FinishedTasksExport = ({tasks}) => {
    const headers = ['ID', 'TITLE', 'DESCRIPTION', 'START TIME', 'END TIME', 'DURATION'];
    const csvData = [headers];
    tasks.forEach( (task) => {
        const start_moment = moment(task.start_date);
        const end_moment = moment(task.end_date);
        const row = [ 
            task.id, 
            task.title, 
            task.description, 
            start_moment.format('hh:mm:ss A'), 
            end_moment.format('hh:mm:ss A'),
            `${end_moment.diff(start_moment, 'minutes')}m`
        ]
        csvData.push(row);
    });
    const printDate = moment(tasks[0].start_date).format('YYYYMMDD');
    const fileName = `pomodoro-app-export-${printDate}.csv`
    return (
        <CSVLink data={csvData} filename={fileName} enclosingCharacter={`"`} className="finished-tasks__link">
            <div className="task-form__button finished-tasks__link-export button-wide">Export</div>
        </CSVLink>
    );
};

export default FinishedTasksExport;