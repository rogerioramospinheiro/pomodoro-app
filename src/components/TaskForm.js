import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { startTask, cancelTask } from '../actions/tasks';

class TaskForm extends React.Component {
    state = {
        task_title: '',
        task_description: '',
    };
    onDescriptionChangeHandle = (e) => {
        const description = e.target.value;
        this.setState( {task_description: description} );
    };
    onTitleChangeHandle = (e) => {
        const title = e.target.value;
        this.setState( {task_title: title} );
    };
    toggleButtonName = () => {
        return this.isIdle() ? 'Start' : 'Stop';
    };
    onSubmitHandle = (e) => {
        e.preventDefault();
        if (this.isIdle()) {
            this.props.startTask({
                title: this.state.task_title, 
                description: this.state.task_description, 
                start_date: moment().valueOf(), 
            })
        } else {
            this.props.cancelTask();
        }
    };
    isIdle = () => {
        return this.props.state_name === 'IDLE';
    };
    isNotIdle = () => {
        return this.props.state_name !== 'IDLE';
    }
    render() {
        return (
            <div>
                <form onSubmit={this.onSubmitHandle}>
                    <input type="text"
                    placeholder="Task Title"
                    autoFocus
                    value={this.state.task_title}
                    onChange={this.onTitleChangeHandle}
                    disabled={this.isNotIdle()}
                    />
                    <input type="text"
                    placeholder="Task Description"
                    value={this.state.task_description}
                    onChange={this.onDescriptionChangeHandle}
                    disabled={this.isNotIdle()}
                    />
                    <button>{ this.toggleButtonName() }</button>
                </form>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        state_name: state.tasks.state_name
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        startTask: (task) => dispatch(startTask(task)),
        cancelTask: () => dispatch(cancelTask())
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(TaskForm);