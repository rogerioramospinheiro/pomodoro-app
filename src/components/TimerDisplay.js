import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { finishTask } from '../actions/tasks';
import { selectInitTime, selectDisplayColor } from '../selectors/settings';

class TimerDisplay extends React.Component {
    state = {
        minutes: 0,
        seconds: 0,
    };
    onIntervalHandle = (ctx) => {

        const timer = ctx.currentCounter();

        if (timer) {

            let minutes = timer.minutes;
            let seconds = timer.seconds;
            
            if (minutes > 0 && seconds === 0) {
                minutes = minutes - 1;
                seconds = 59;
            } else {
                seconds = seconds - 1;
            }
    
            if ( minutes === 0 && seconds === 0) {
                ctx.finishCountdown();
            }
            
            ctx.setCounter(minutes, seconds);
        }
    };
    setCounter = (minutes, seconds) => {
        this.setState( {minutes: minutes, seconds: seconds} );
    };
    displayCounter = () => {
        const m = this.state.minutes < 10 ? `0${this.state.minutes}` : this.state.minutes;
        const s = this.state.seconds < 10 ? `0${this.state.seconds}` : this.state.seconds;
        return `${m}:${s}`;
    };
    currentCounter = () => {

        const minutes = this.state.minutes;
        const seconds = this.state.seconds;

        if (this.props.init_time > 0) {

            if (minutes > 0 || seconds > 0) {
                return {
                    minutes: minutes,
                    seconds: seconds
                }
            } else {
                return {
                    minutes: this.props.init_time,
                    seconds: 0
                }
            }

        } else {

            if (minutes > 0 || seconds > 0) {
                this.setCounter(0, 0);
            }

            return undefined;
        }
    };
    finishCountdown = () => {
        const now = {
            end_date: moment().valueOf()
        }
        this.props.finishTask(now);
    };
    componentDidMount() {
        this.intervalTimeout = setInterval(this.onIntervalHandle, 1000, this);
    }
    componentWillUnmount() {
        clearInterval(this.intervalTimeout);
    }
    render() {
        return (
            <div className="timer-display">
                <div className="timer-display__text" style={{color: this.props.display_color}}>{ this.displayCounter() }</div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        init_time: selectInitTime(state.tasks, state.settings),
        display_color: selectDisplayColor(state.tasks)
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        finishTask: (end_date) => dispatch(finishTask(end_date))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(TimerDisplay);