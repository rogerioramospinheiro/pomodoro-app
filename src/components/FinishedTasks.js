import React from 'react';
import { connect } from 'react-redux';
import FinishedTaskItem from './FinishedTaskItem';

export const FinishedTasks = (props) => (
    <div className="finished-tasks" >
        <FinishedTasksList {...props} />
    </div>
);

export const FinishedTasksList = (props) => {
    if (props.tasks.length === 0) {
        return (<div className="finished-tasks__item">No finished tasks</div>);
    } else {
        return props.tasks.map( forEachTask )
    }
}

const forEachTask = (task, index) => (
    <FinishedTaskItem key={index} {...task}/>
);

const mapStateToProps = (state) => {
    return {
        tasks: state.tasks.finished_tasks
    };
};

export default connect(mapStateToProps)(FinishedTasks);