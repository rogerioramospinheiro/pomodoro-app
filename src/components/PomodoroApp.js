import React from 'react';
import TaskForm from './TaskForm';
import TimerDisplay from './TimerDisplay';
import FinishedTasks from './FinishedTasks';

const PomodoroApp = () => (
    <div>
        <TimerDisplay />
        <TaskForm />
        <FinishedTasks />
    </div>
);

export default PomodoroApp;