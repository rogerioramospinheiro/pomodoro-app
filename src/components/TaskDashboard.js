import React from 'react';
import TaskForm from './TaskForm';
import TimerDisplay from './TimerDisplay';
import FinishedTasks from './FinishedTasks';

const TaskDashboard = () => (
    <div>
        <TimerDisplay />
        <TaskForm />
        <FinishedTasks />
    </div>
);

export default TaskDashboard;