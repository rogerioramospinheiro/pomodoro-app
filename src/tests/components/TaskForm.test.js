import React from 'react';
import { shallow } from 'enzyme';
import { TaskForm } from '../../components/TaskForm';

jest.mock('moment', () => {
    return () => ({
        valueOf: function() {
            return 1000;
        }
    });
});

const createIdleTaskForm = () => {
    const state_name = 'IDLE';
    const wrapper = shallow(<TaskForm state_name={state_name} />);
    return wrapper;
};

const createIdleTaskFormAndRunSubmit = ({task_title = '', task_description = ''} = {}) => {
    // Given
    const state_name = 'IDLE';
    const startTask = jest.fn();
    const wrapper = shallow(<TaskForm state_name={state_name} startTask={startTask}/>);
    wrapper.setState({
        task_title, 
        task_description
    });

    // When
    wrapper.find('form').simulate('submit', {
        preventDefault: jest.fn()
    });

    return startTask;
};

test('Should render TaskForm in idle state', () => {
    // Given
    const wrapper = createIdleTaskForm();
    // Then
    expect(wrapper).toMatchSnapshot();
});

test('Should render TaskForm in counting state', () => {
    // Given
    const state_name = 'COUNTING';
    const wrapper = shallow(<TaskForm state_name={state_name} />);
    wrapper.setState({
        task_title: 'JOB-0005', 
        task_description: 'Counting...'
    });

    // Then
    expect(wrapper).toMatchSnapshot();
});

test('Should handle onDescriptionChangeHandle', () => {
    // Given
    const wrapper = createIdleTaskForm();
    const value = 'JOB-0001';
    // When
    wrapper.find('input').simulate('change', {
        target: {
            value
        }
    });
    // Then
    expect(wrapper.state('task_title')).toBe('JOB-0001');
});

test('Should handle onTitleChangeHandle', () => {
    // Given
    const wrapper = createIdleTaskForm();
    const value = 'First step to finish the job';
    // When
    wrapper.find('textarea').simulate('change', {
        target: {
            value
        }
    });
    // Then
    expect(wrapper.state('task_description')).toBe('First step to finish the job');
});

test('Should handle onSubmit and startTask', () => {
    // Given
    const startTask = createIdleTaskFormAndRunSubmit({
        task_title: 'JOB-0001',
        task_description: 'Second attempt to finish the job'
    });

    // Then
    expect(startTask.mock.calls.length).toBe(1);
    expect(startTask.mock.calls[0][0]).toEqual({
        title: 'JOB-0001',
        description: 'Second attempt to finish the job',
        start_date: 1000
    });
});

test('Should handle onSubmit and not startTask missing description', () => {
    // Given
    const startTask = createIdleTaskFormAndRunSubmit({
        task_title: 'JOB-0001'
    });

    // Then
    expect(startTask.mock.calls.length).toBe(0);
});

test('Should handle onSubmit and not startTask missing title', () => {
    // Given
    const startTask = createIdleTaskFormAndRunSubmit({
        description: 'Second attempt to finish the job'
    });

    // Then
    expect(startTask.mock.calls.length).toBe(0);
});

test('Should handle onSubmit and not startTask missing title and description', () => {
    // Given
    const startTask = createIdleTaskFormAndRunSubmit();

    // Then
    expect(startTask.mock.calls.length).toBe(0);
});

test('Should handle onSubmit and cancelTask', () => {
    // Given
    const state_name = 'COUNTING';
    const cancelTask = jest.fn();
    const wrapper = shallow(<TaskForm state_name={state_name} cancelTask={cancelTask}/>);
    wrapper.setState({
        task_title: 'JOB-0003', 
        task_description: 'Canceling...'
    });

    // When
    wrapper.find('form').simulate('submit', {
        preventDefault: jest.fn()
    });

    // Then
    expect(cancelTask.mock.calls.length).toBe(1);
});

test('Should handle onFinishHandle', () => {
    // Given
    const state_name = 'COUNTING';
    const terminateTask = jest.fn();
    const wrapper = shallow(<TaskForm state_name={state_name} terminateTask={terminateTask}/>);
    wrapper.setState({
        task_title: 'JOB-0004', 
        task_description: 'Terminating...'
    });

    // When
    wrapper.find('button').at(1).simulate('click', {
        preventDefault: jest.fn()
    });

    // Then
    expect(terminateTask.mock.calls[0][0]).toEqual({
        end_date: 1000
    });
});