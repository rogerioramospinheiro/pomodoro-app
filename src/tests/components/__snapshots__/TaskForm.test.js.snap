// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render TaskForm in counting state 1`] = `
<div
  className="task-form"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="task-form__spacebox"
    >
      <input
        autoFocus={true}
        className="task-form__input"
        disabled={true}
        onChange={[Function]}
        placeholder="Task Title"
        type="text"
        value="JOB-0005"
      />
    </div>
    <div
      className="task-form__spacebox"
    >
      <textarea
        className="task-form__textarea"
        disabled={true}
        onChange={[Function]}
        placeholder="Task Description"
        type="text"
        value="Counting..."
      />
    </div>
    <div
      className="task-form__button-group"
    >
      <button
        className="task-form__button button-short"
      >
        Cancel
      </button>
      <button
        className="task-form__button button-short"
        onClick={[Function]}
      >
        Finish
      </button>
    </div>
  </form>
</div>
`;

exports[`Should render TaskForm in idle state 1`] = `
<div
  className="task-form"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="task-form__spacebox"
    >
      <input
        autoFocus={true}
        className="task-form__input"
        disabled={false}
        onChange={[Function]}
        placeholder="Task Title"
        type="text"
        value=""
      />
    </div>
    <div
      className="task-form__spacebox"
    >
      <textarea
        className="task-form__textarea"
        disabled={false}
        onChange={[Function]}
        placeholder="Task Description"
        type="text"
        value=""
      />
    </div>
    <button
      className="task-form__button button-wide"
    >
      Start
    </button>
  </form>
</div>
`;
